subsys/bluetooth/services/ots/ots_srv_oacp.c

```diff
@@ -0,0 +1,344 @@
+		break;
+	case BT_GATT_OTS_SRV_OACP_PROC_ABORT:
+	default:
+		break;
+	}
+
+	LOG_WRN("OACP unsupported procedure type");
+	return BT_GATT_OTS_SRV_OACP_RES_OPCODE_NOT_SUP;
+}
+
+static bool oacp_command_len_verify(struct bt_gatt_ots_srv_oacp_proc *proc,
+				    uint16_t len)
+{
+	uint16_t ref_len = OACP_PROC_TYPE_SIZE;
+
+	switch (proc->type) {
+	case BT_GATT_OTS_SRV_OACP_PROC_CREATE:
+	{
+		struct bt_gatt_ots_srv_obj_type *type;
+
+		ref_len += sizeof(proc->create_params.size);
+
+		type = &proc->create_params.type;
+		if (type->uuid.type == BT_UUID_TYPE_16) {
+			ref_len += sizeof(type->uuid_16.val);
+		} else if (type->uuid.type == BT_UUID_TYPE_128) {
+			ref_len += sizeof(type->uuid_128.val);
+		} else {
+			return true;
+		}
+	} break;
+	case BT_GATT_OTS_SRV_OACP_PROC_DELETE:
+		break;
+	case BT_GATT_OTS_SRV_OACP_PROC_CHECKSUM_CALC:
+		ref_len += sizeof(proc->cs_calc_params);
+		break;
+	case BT_GATT_OTS_SRV_OACP_PROC_EXECUTE:
+		break;
+	case BT_GATT_OTS_SRV_OACP_PROC_READ:
+		ref_len += sizeof(proc->read_params);
+		break;
+	case BT_GATT_OTS_SRV_OACP_PROC_WRITE:
+		ref_len += sizeof(proc->write_params);
+		break;
+	case BT_GATT_OTS_SRV_OACP_PROC_ABORT:
+		break;
+	default:
+		return true;
+	}
+
+	return (len == ref_len);
+}
+
+static void oacp_read_proc_cb(struct bt_gatt_ots_srv_l2cap *l2cap_ctx,
+			      struct bt_conn *conn)
+{
+	int err;
+	uint8_t *obj_chunk;
+	uint32_t offset;
+	uint32_t len;
+	struct bt_gatt_ots_srv *ots;
+	struct bt_gatt_ots_srv_object_read_op *read_op;
+
+	ots     = CONTAINER_OF(l2cap_ctx, struct bt_gatt_ots_srv, l2cap);
+	read_op = &ots->cur_obj->state.read_op;
+	offset  = read_op->oacp_params.offset + read_op->sent_len;
+
+	if (read_op->sent_len >= read_op->oacp_params.len) {
+		LOG_DBG("OACP Read Op over L2CAP is completed");
+
+		if (read_op->sent_len > read_op->oacp_params.len) {
+			LOG_WRN("More bytes sent that the client requested");
+		}
+
+		ots->cur_obj->state.type = BT_GATT_OTS_SRV_OBJECT_IDLE_STATE;
+		ots->cb->obj_read(ots, conn, ots->cur_obj->id, NULL, 0,
+				  offset);
+		return;
+	}
+
+	len = read_op->oacp_params.len - read_op->sent_len;
+	len = ots->cb->obj_read(ots, conn, ots->cur_obj->id, &obj_chunk, len,
+				offset);
+
+	ots->l2cap.tx_done = oacp_read_proc_cb;
+	err = bt_gatt_ots_srv_l2cap_send(&ots->l2cap, obj_chunk, len);
+	if (err) {
+		LOG_ERR("L2CAP CoC error: %d while trying to execute OACP "
+			"Read procedure", err);
+		ots->cur_obj->state.type = BT_GATT_OTS_SRV_OBJECT_IDLE_STATE;
+	} else {
+		read_op->sent_len += len;
+	}
+}
+
+static void oacp_read_proc_execute(struct bt_gatt_ots_srv *ots,
+				   struct bt_conn *conn)
+{
+	struct bt_gatt_ots_srv_oacp_read_params *params =
+		&ots->cur_obj->state.read_op.oacp_params;
+
+	if (!ots->cur_obj) {
+		LOG_ERR("Invalid Current Object on OACP Read procedure");
+		return;
+	}
+
+	LOG_DBG("Executing Read procedure with offset: 0x%08X and "
+		"length: 0x%08X", params->offset, params->len);
+
+	if (ots->cb->obj_read) {
```

Comment:
Didn't you add an __ASSERT for this? I guess the check can be removed, unless the client of course changes the callback during runtime...

Reply:
Well, I realized that supporting Object read operation is optional so I guess I will leave it here and correct my ASSERT logic.